### DISCUSSION TITLE
Difference between CldUploadButton and CldUploadWidget

### DISCUSSION THREAD

**Discussion by roopeshsn** [cloudinary-community/next-cloudinary#86] [General]
Looks like the components CldUploadButton and CldUploadWidget are doing the same work. Both opening the widget. Why is that? or anything I am missing out on?

**Comment by colbyfayock** [comment]
They're almost the same. CldUploadButton gives you a pre-made button that opens the widget with less options, where CldUploadWidget requires you to manually add the button, enabling more customization

intent is a simple option and a more configurable option

**Reply by destino92** [reply]
How can i check that cloudinary is loaded before mounting the CldUploadButton? The user might click on the button before the script is loaded and i get an error

**Reply by colbyfayock** [reply]
hey @destino92 the Cloudinary scripts load as part of the CldUploadButton component, if you dont mount the component, it's not going to ever tell you it's loaded

what error are you getting? when the open event is triggered, it tries to create a new widget, but if the Cloudinary script still isn't loaded at that point, yeah i would imagine it would fail

there are probably a few ways i could handle this in the library, im thinking the best way may be to pass the loading state down into the Button and disable the button until it's available. thatd likely be the least hacky solution. the only issue would be for CldUploadWidget someone would need to implement that themselves, which isn't a big deal, but should be aware
